src/Exchange/SwapExchange.php

```diff
@@ -18,28 +18,38 @@
 final class SwapExchange implements Exchange
 {
     /**
-     * @var SwapInterface
+     * @var Swap
      */
     private $swap;
 
     /**
-     * @param SwapInterface $swap
+     * @param Swap|ExchangeRateProvider $exchange
      */
-    public function __construct(SwapInterface $swap)
+    public function __construct($exchange)
     {
-        $this->swap = $swap;
+        if (!$exchange instanceof Swap && !$exchange instanceof ExchangeRateProvider) {
+            throw new \InvalidArgumentException(sprintf(
+                'Exchange must either be %s or %s',
+                Swap::class,
+                ExchangeRateProvider::class
+            ));
+        }
+
+        if ($exchange instanceof ExchangeRateProvider) {
+            $exchange = new Swap($exchange);
```

Comment:
Anyway, what about just not accepting `ExchangeRateProvider` here, it seems easy to make `Swap` instance out of it and will reduce complexity.